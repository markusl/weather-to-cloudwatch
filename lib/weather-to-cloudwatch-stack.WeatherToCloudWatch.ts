import { CloudWatch } from '@aws-sdk/client-cloudwatch';

const cw = new CloudWatch({});

interface WeatherData {
	'coord': {
		'lon': number;
		'lat': number;
	},
	'weather': [{
		'id': number;
		'main': string;
		'description': string;
		'icon': string;
	}],
	'base': string;
	'main': {
		'temp': number;
		'feels_like': number;
		'temp_min': number;
		'temp_max': number;
		'pressure': number;
		'humidity': number;
	},
	'visibility': number;
	'wind': {
		'speed': number;
		'deg': number;
	},
	'clouds': {
		'all': number;
	},
	'dt': number;
	'sys': {
		'type': number;
		'id': number;
		'country': string;
		'sunrise': number;
		'sunset': number;
	},
	'timezone': number;
	'id': number;
	'name': string;
	'cod': number;
}

export const handler = async (event: any) => {
  console.log(JSON.stringify(event, undefined, 2));
  try {
    const appId = process.env.OPEN_WEATHER_APP_ID;
    const weatherUrl = `https://api.openweathermap.org/data/2.5/weather?lat=60.192059&lon=24.945831&appid=${appId}&units=metric`;
    const weatherData = await (await fetch(weatherUrl)).json() as WeatherData;
    console.log(weatherData);

    await cw.putMetricData({
      Namespace: 'Weather',
      MetricData: [
        {
          MetricName: 'Temperature',
          Dimensions: [{
            Name: 'City',
            Value: 'Helsinki',
          }],
          Value: weatherData.main.temp,
        },
        {
          MetricName: 'Humidity',
          Dimensions: [{
            Name: 'City',
            Value: 'Helsinki',
          }],
          Value: weatherData.main.humidity,
        },
        {
          MetricName: 'Pressure',
          Dimensions: [{
            Name: 'City',
            Value: 'Helsinki',
          }],
          Value: weatherData.main.pressure,
        },
      ]
    });

    return {
      statusCode: 200,
    };
  } catch (e) {
    console.error(e);
    return {
      statusCode: 500,
      body: JSON.stringify(e)
    };
  }
};
