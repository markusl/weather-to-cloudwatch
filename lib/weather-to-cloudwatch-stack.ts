import { Construct } from 'constructs';
import * as cdk from 'aws-cdk-lib';
import {
  aws_cloudwatch as cloudwatch,
  aws_lambda as lambda,
  aws_lambda_nodejs as lambda_nodejs,
  aws_events as events,
  aws_events_targets as targets,
  aws_ssm as ssm,
  pipelines,
} from 'aws-cdk-lib';

const everyFiveMinutes = events.Schedule.expression('cron(0/5 * * * ? *)');

export class WeatherToCloudWatch extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Use a connection created using the AWS console to authenticate to GitHub
    const input = pipelines.CodePipelineSource.connection('markusl/weather-to-cloudwatch', 'main', {
      connectionArn: `arn:aws:codestar-connections:${cdk.Stack.of(this).region}:${cdk.Stack.of(this).account}:connection/2176abff-fac4-4c5d-87e8-0cc53551ab98`,
    });

    new pipelines.CodePipeline(this, 'WeatherToCloudWatchPipeline', {
      crossAccountKeys: false,
      synth: new pipelines.ShellStep('Synth', {
        input,
        commands: [
          'npm ci',
          'npm run test',
          'npx cdk synth',
        ],
      }),
    });

    const updateWeatherFunction = new lambda_nodejs.NodejsFunction(this, 'WeatherToCloudWatch', {
      runtime: lambda.Runtime.NODEJS_18_X,
      bundling: {
        minify: true,
        externalModules: [
          '@aws-sdk/client-s3',
          '@aws-sdk/client-cloudwatch',
        ],
      },
      memorySize: 256,
      timeout: cdk.Duration.minutes(1),
      architecture: lambda.Architecture.ARM_64,
      environment: {
        OPEN_WEATHER_APP_ID: ssm.StringParameter.fromStringParameterName(this, 'AppId', 'OPEN_WEATHER_APP_ID').stringValue
      },
    });

    cloudwatch.Metric.grantPutMetricData(updateWeatherFunction);

    const rule = new events.Rule(this, 'Rule', { schedule: everyFiveMinutes });
    rule.addTarget(new targets.LambdaFunction(updateWeatherFunction));
  }
}
