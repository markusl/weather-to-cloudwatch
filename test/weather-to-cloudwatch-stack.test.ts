import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { WeatherToCloudWatch } from '../lib/weather-to-cloudwatch-stack';
import { snapshotSerializer } from './util';

expect.addSnapshotSerializer(snapshotSerializer);

test('WeatherToCloudWatch synthetizes expected template', async () => {
  const app = new cdk.App();
  const stack = new WeatherToCloudWatch(app, 'infra');

  expect(Template.fromStack(stack)).toMatchSnapshot();
});
