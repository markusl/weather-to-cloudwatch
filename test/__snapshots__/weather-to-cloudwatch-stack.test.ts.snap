// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherToCloudWatch synthetizes expected template 1`] = `
{
  "Parameters": {
    "AppIdParameter": {
      "Default": "OPEN_WEATHER_APP_ID",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "Rule4C995B7F": {
      "Properties": {
        "ScheduleExpression": "cron(0/5 * * * ? *)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "WeatherToCloudWatchA852C2DE",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "RuleAllowEventRuleinfraWeatherToCloudWatch246E2F9276CFBCBB": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "WeatherToCloudWatchA852C2DE",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "Rule4C995B7F",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "WeatherToCloudWatchA852C2DE": {
      "DependsOn": [
        "WeatherToCloudWatchServiceRoleDefaultPolicy9921346E",
        "WeatherToCloudWatchServiceRoleBA4410DE",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "[HASH REMOVED].zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "OPEN_WEATHER_APP_ID": {
              "Ref": "AppIdParameter",
            },
          },
        },
        "Handler": "index.handler",
        "MemorySize": 256,
        "Role": {
          "Fn::GetAtt": [
            "WeatherToCloudWatchServiceRoleBA4410DE",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "WeatherToCloudWatchPipelineArtifactsBucketAF443023": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "SSEAlgorithm": "aws:kms",
              },
            },
          ],
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "WeatherToCloudWatchPipelineArtifactsBucketPolicyD096D2EF": {
      "Properties": {
        "Bucket": {
          "Ref": "WeatherToCloudWatchPipelineArtifactsBucketAF443023",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "WeatherToCloudWatchPipelineArtifactsBucketAF443023",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "WeatherToCloudWatchPipelineArtifactsBucketAF443023",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "WeatherToCloudWatchPipelineBuildSynthCdkBuildProject4B598C0F": {
      "Properties": {
        "Artifacts": {
          "Type": "CODEPIPELINE",
        },
        "Cache": {
          "Type": "NO_CACHE",
        },
        "Description": "Pipeline step infra/Pipeline/Build/Synth",
        "EncryptionKey": "alias/aws/s3",
        "Environment": {
          "ComputeType": "BUILD_GENERAL1_SMALL",
          "Image": "aws/codebuild/standard:5.0",
          "ImagePullCredentialsType": "CODEBUILD",
          "PrivilegedMode": false,
          "Type": "LINUX_CONTAINER",
        },
        "ServiceRole": {
          "Fn::GetAtt": [
            "WeatherToCloudWatchPipelineBuildSynthCdkBuildProjectRole1BC357A7",
            "Arn",
          ],
        },
        "Source": {
          "BuildSpec": "{
  "version": "0.2",
  "phases": {
    "build": {
      "commands": [
        "npm ci",
        "npm run test",
        "npx cdk synth"
      ]
    }
  },
  "artifacts": {
    "base-directory": "cdk.out",
    "files": "**/*"
  }
}",
          "Type": "CODEPIPELINE",
        },
      },
      "Type": "AWS::CodeBuild::Project",
    },
    "WeatherToCloudWatchPipelineBuildSynthCdkBuildProjectRole1BC357A7": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "codebuild.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "WeatherToCloudWatchPipelineBuildSynthCdkBuildProjectRoleDefaultPolicy5FBD768A": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:",
                      {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":log-group:/aws/codebuild/",
                      {
                        "Ref": "WeatherToCloudWatchPipelineBuildSynthCdkBuildProject4B598C0F",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:",
                      {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":log-group:/aws/codebuild/",
                      {
                        "Ref": "WeatherToCloudWatchPipelineBuildSynthCdkBuildProject4B598C0F",
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "codebuild:CreateReportGroup",
                "codebuild:CreateReport",
                "codebuild:UpdateReport",
                "codebuild:BatchPutTestCases",
                "codebuild:BatchPutCodeCoverages",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":codebuild:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":report-group/",
                    {
                      "Ref": "WeatherToCloudWatchPipelineBuildSynthCdkBuildProject4B598C0F",
                    },
                    "-*",
                  ],
                ],
              },
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "WeatherToCloudWatchPipelineArtifactsBucketAF443023",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "WeatherToCloudWatchPipelineArtifactsBucketAF443023",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "WeatherToCloudWatchPipelineBuildSynthCdkBuildProjectRoleDefaultPolicy5FBD768A",
        "Roles": [
          {
            "Ref": "WeatherToCloudWatchPipelineBuildSynthCdkBuildProjectRole1BC357A7",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "WeatherToCloudWatchPipelineC815E69C": {
      "DependsOn": [
        "WeatherToCloudWatchPipelineRoleDefaultPolicy74D70E78",
        "WeatherToCloudWatchPipelineRoleE9DDE380",
      ],
      "Properties": {
        "ArtifactStore": {
          "Location": {
            "Ref": "WeatherToCloudWatchPipelineArtifactsBucketAF443023",
          },
          "Type": "S3",
        },
        "RestartExecutionOnUpdate": true,
        "RoleArn": {
          "Fn::GetAtt": [
            "WeatherToCloudWatchPipelineRoleE9DDE380",
            "Arn",
          ],
        },
        "Stages": [
          {
            "Actions": [
              {
                "ActionTypeId": {
                  "Category": "Source",
                  "Owner": "AWS",
                  "Provider": "CodeStarSourceConnection",
                  "Version": "1",
                },
                "Configuration": {
                  "BranchName": "main",
                  "ConnectionArn": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:aws:codestar-connections:",
                        {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":connection/2176abff-fac4-4c5d-87e8-0cc53551ab98",
                      ],
                    ],
                  },
                  "FullRepositoryId": "markusl/weather-to-cloudwatch",
                },
                "Name": "markusl_weather-to-cloudwatch",
                "OutputArtifacts": [
                  {
                    "Name": "markusl_weather_to_cloudwatch_Source",
                  },
                ],
                "RoleArn": {
                  "Fn::GetAtt": [
                    "WeatherToCloudWatchPipelineSourcemarkuslweathertocloudwatchCodePipelineActionRoleAEED61F1",
                    "Arn",
                  ],
                },
                "RunOrder": 1,
              },
            ],
            "Name": "Source",
          },
          {
            "Actions": [
              {
                "ActionTypeId": {
                  "Category": "Build",
                  "Owner": "AWS",
                  "Provider": "CodeBuild",
                  "Version": "1",
                },
                "Configuration": {
                  "EnvironmentVariables": "[{"name":"_PROJECT_CONFIG_HASH","type":"PLAINTEXT","value":"e0f4822031066107e1c3913016095aaedfa15bc7cea15d7405133bc5d55d02d3"}]",
                  "ProjectName": {
                    "Ref": "WeatherToCloudWatchPipelineBuildSynthCdkBuildProject4B598C0F",
                  },
                },
                "InputArtifacts": [
                  {
                    "Name": "markusl_weather_to_cloudwatch_Source",
                  },
                ],
                "Name": "Synth",
                "OutputArtifacts": [
                  {
                    "Name": "Synth_Output",
                  },
                ],
                "RoleArn": {
                  "Fn::GetAtt": [
                    "WeatherToCloudWatchPipelineCodeBuildActionRole37996480",
                    "Arn",
                  ],
                },
                "RunOrder": 1,
              },
            ],
            "Name": "Build",
          },
          {
            "Actions": [
              {
                "ActionTypeId": {
                  "Category": "Build",
                  "Owner": "AWS",
                  "Provider": "CodeBuild",
                  "Version": "1",
                },
                "Configuration": {
                  "EnvironmentVariables": "[{"name":"_PROJECT_CONFIG_HASH","type":"PLAINTEXT","value":"c302495089e029880b6057af1b45e7ba2ecd996ed4944c2a9238447daed4a388"}]",
                  "ProjectName": {
                    "Ref": "WeatherToCloudWatchPipelineUpdatePipelineSelfMutation1A8D496A",
                  },
                },
                "InputArtifacts": [
                  {
                    "Name": "Synth_Output",
                  },
                ],
                "Name": "SelfMutate",
                "RoleArn": {
                  "Fn::GetAtt": [
                    "WeatherToCloudWatchPipelineCodeBuildActionRole37996480",
                    "Arn",
                  ],
                },
                "RunOrder": 1,
              },
            ],
            "Name": "UpdatePipeline",
          },
        ],
      },
      "Type": "AWS::CodePipeline::Pipeline",
    },
    "WeatherToCloudWatchPipelineCodeBuildActionRole37996480": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Condition": {
                "Bool": {
                  "aws:ViaAWSService": "codepipeline.amazonaws.com",
                },
              },
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "WeatherToCloudWatchPipelineCodeBuildActionRoleDefaultPolicy91BA76AA": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "codebuild:BatchGetBuilds",
                "codebuild:StartBuild",
                "codebuild:StopBuild",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "WeatherToCloudWatchPipelineBuildSynthCdkBuildProject4B598C0F",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "codebuild:BatchGetBuilds",
                "codebuild:StartBuild",
                "codebuild:StopBuild",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "WeatherToCloudWatchPipelineUpdatePipelineSelfMutation1A8D496A",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "WeatherToCloudWatchPipelineCodeBuildActionRoleDefaultPolicy91BA76AA",
        "Roles": [
          {
            "Ref": "WeatherToCloudWatchPipelineCodeBuildActionRole37996480",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "WeatherToCloudWatchPipelineRoleDefaultPolicy74D70E78": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "WeatherToCloudWatchPipelineArtifactsBucketAF443023",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "WeatherToCloudWatchPipelineArtifactsBucketAF443023",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "WeatherToCloudWatchPipelineSourcemarkuslweathertocloudwatchCodePipelineActionRoleAEED61F1",
                  "Arn",
                ],
              },
            },
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "WeatherToCloudWatchPipelineCodeBuildActionRole37996480",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "WeatherToCloudWatchPipelineRoleDefaultPolicy74D70E78",
        "Roles": [
          {
            "Ref": "WeatherToCloudWatchPipelineRoleE9DDE380",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "WeatherToCloudWatchPipelineRoleE9DDE380": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "codepipeline.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "WeatherToCloudWatchPipelineSourcemarkuslweathertocloudwatchCodePipelineActionRoleAEED61F1": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "WeatherToCloudWatchPipelineSourcemarkuslweathertocloudwatchCodePipelineActionRoleDefaultPolicyB8CAB876": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "codestar-connections:UseConnection",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:codestar-connections:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":connection/2176abff-fac4-4c5d-87e8-0cc53551ab98",
                  ],
                ],
              },
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "WeatherToCloudWatchPipelineArtifactsBucketAF443023",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "WeatherToCloudWatchPipelineArtifactsBucketAF443023",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:PutObjectAcl",
                "s3:PutObjectVersionAcl",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "WeatherToCloudWatchPipelineArtifactsBucketAF443023",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "WeatherToCloudWatchPipelineSourcemarkuslweathertocloudwatchCodePipelineActionRoleDefaultPolicyB8CAB876",
        "Roles": [
          {
            "Ref": "WeatherToCloudWatchPipelineSourcemarkuslweathertocloudwatchCodePipelineActionRoleAEED61F1",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "WeatherToCloudWatchPipelineUpdatePipelineSelfMutation1A8D496A": {
      "Properties": {
        "Artifacts": {
          "Type": "CODEPIPELINE",
        },
        "Cache": {
          "Type": "NO_CACHE",
        },
        "Description": "Pipeline step infra/Pipeline/UpdatePipeline/SelfMutate",
        "EncryptionKey": "alias/aws/s3",
        "Environment": {
          "ComputeType": "BUILD_GENERAL1_SMALL",
          "Image": "aws/codebuild/standard:5.0",
          "ImagePullCredentialsType": "CODEBUILD",
          "PrivilegedMode": false,
          "Type": "LINUX_CONTAINER",
        },
        "ServiceRole": {
          "Fn::GetAtt": [
            "WeatherToCloudWatchPipelineUpdatePipelineSelfMutationRole70B7DC09",
            "Arn",
          ],
        },
        "Source": {
          "BuildSpec": "{
  "version": "0.2",
  "phases": {
    "install": {
      "commands": [
        "npm install -g aws-cdk@2"
      ]
    },
    "build": {
      "commands": [
        "cdk -a . deploy infra --require-approval=never --verbose"
      ]
    }
  }
}",
          "Type": "CODEPIPELINE",
        },
      },
      "Type": "AWS::CodeBuild::Project",
    },
    "WeatherToCloudWatchPipelineUpdatePipelineSelfMutationRole70B7DC09": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "codebuild.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "WeatherToCloudWatchPipelineUpdatePipelineSelfMutationRoleDefaultPolicy76F9E2D3": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:",
                      {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":log-group:/aws/codebuild/",
                      {
                        "Ref": "WeatherToCloudWatchPipelineUpdatePipelineSelfMutation1A8D496A",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:",
                      {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":log-group:/aws/codebuild/",
                      {
                        "Ref": "WeatherToCloudWatchPipelineUpdatePipelineSelfMutation1A8D496A",
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "codebuild:CreateReportGroup",
                "codebuild:CreateReport",
                "codebuild:UpdateReport",
                "codebuild:BatchPutTestCases",
                "codebuild:BatchPutCodeCoverages",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":codebuild:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":report-group/",
                    {
                      "Ref": "WeatherToCloudWatchPipelineUpdatePipelineSelfMutation1A8D496A",
                    },
                    "-*",
                  ],
                ],
              },
            },
            {
              "Action": "sts:AssumeRole",
              "Condition": {
                "ForAnyValue:StringEquals": {
                  "iam:ResourceTag/aws-cdk:bootstrap-role": [
                    "image-publishing",
                    "file-publishing",
                    "deploy",
                  ],
                },
              },
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:*:iam::",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":role/*",
                  ],
                ],
              },
            },
            {
              "Action": "cloudformation:DescribeStacks",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "s3:ListBucket",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "WeatherToCloudWatchPipelineArtifactsBucketAF443023",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "WeatherToCloudWatchPipelineArtifactsBucketAF443023",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "WeatherToCloudWatchPipelineUpdatePipelineSelfMutationRoleDefaultPolicy76F9E2D3",
        "Roles": [
          {
            "Ref": "WeatherToCloudWatchPipelineUpdatePipelineSelfMutationRole70B7DC09",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "WeatherToCloudWatchServiceRoleBA4410DE": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "WeatherToCloudWatchServiceRoleDefaultPolicy9921346E": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "cloudwatch:PutMetricData",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "WeatherToCloudWatchServiceRoleDefaultPolicy9921346E",
        "Roles": [
          {
            "Ref": "WeatherToCloudWatchServiceRoleBA4410DE",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
